processor/src/fast/mod.rs

```diff
@@ -0,0 +1,881 @@
+        program: &MastForest,
+        kernel: &Kernel,
+        host: &mut impl Host,
+    ) -> Result<(), ExecutionError> {
+        // dyn calls are not allowed inside a syscall
+        if dyn_node.is_dyncall() && self.in_syscall {
+            return Err(ExecutionError::CallInSyscall("dyncall"));
+        }
+
+        // Retrieve callee hash from memory, using stack top as the memory address.
+        let callee_hash = {
+            let mem_addr = self.stack[self.stack_top_idx - 1];
+            self.memory.read_word(self.ctx, mem_addr, self.clk).copied()?
+        };
+
+        // Drop the memory address from the stack. This needs to be done BEFORE saving the context,
+        // because the next instruction starts with a "shifted left" stack.
+        self.decrement_stack_size();
+
+        // For dyncall, save the context and reset it.
+        if dyn_node.is_dyncall() {
+            self.save_context_and_truncate_stack();
+            self.ctx = self.clk.into();
+            self.fmp = Felt::new(FMP_MIN);
+            self.fn_hash = callee_hash;
+        };
+
+        // if the callee is not in the program's MAST forest, try to find a MAST forest for it in
+        // the host (corresponding to an external library loaded in the host); if none are
+        // found, return an error.
+        match program.find_procedure_root(callee_hash.into()) {
+            Some(callee_id) => self.execute_mast_node(callee_id, program, kernel, host)?,
+            None => {
+                let mast_forest = host
+                    .get_mast_forest(&callee_hash.into())
+                    .ok_or_else(|| ExecutionError::DynamicNodeNotFound(callee_hash.into()))?;
+
+                // We limit the parts of the program that can be called externally to procedure
+                // roots, even though MAST doesn't have that restriction.
+                let root_id = mast_forest.find_procedure_root(callee_hash.into()).ok_or(
+                    ExecutionError::MalformedMastForestInHost { root_digest: callee_hash.into() },
+                )?;
+
+                self.execute_mast_node(root_id, &mast_forest, kernel, host)?
+            },
+        }
+
+        // For dyncall, restore the context.
+        if dyn_node.is_dyncall() {
+            self.restore_context()?;
+        }
+
+        Ok(())
+    }
+
+    fn execute_external_node(
+        &mut self,
+        external_node: &ExternalNode,
+        kernel: &Kernel,
+        host: &mut impl Host,
+    ) -> Result<(), ExecutionError> {
+        let (root_id, mast_forest) = resolve_external_node(external_node, host)?;
+
+        self.execute_mast_node(root_id, &mast_forest, kernel, host)
+    }
+
+    // Note: when executing individual ops, we do not increment the clock by 1 at every iteration
+    // for performance reasons (~25% performance drop). Hence, `self.clk` cannot be used directly to
+    // determine the number of operations executed in a program.
+    fn execute_basic_block_node(
```

Comment:
Note: in order to compute the `clk` properly, I had to bring in the complex operation batch handling logic from `Process`. This results in a pretty significant performance drop - `blake3_2to1` is now running at around 210 MHz.

Reply:
To fix this, we may need to refactor how the `BasicBlockNode` is structured (i.e., to make iterating over operations as simple as possible).

Reply:
Added to #1701